pkg/controllerhelpers/wait.go

```diff
@@ -31,6 +34,73 @@ type WaitForStateOptions struct {
 	TolerateDelete bool
 }
 
+type AggregatedConditions[Obj runtime.Object] struct {
+	conditions []func(obj Obj) (bool, error)
+	state      []*bool
+}
+
+func NewAggregatedConditions[Obj runtime.Object](condition func(obj Obj) (bool, error), additionalConditions ...func(obj Obj) (bool, error)) *AggregatedConditions[Obj] {
+	conditions := make([]func(obj Obj) (bool, error), 0, 1+len(additionalConditions))
+	conditions = append(conditions, condition)
+	if len(additionalConditions) != 0 {
+		conditions = append(conditions, additionalConditions...)
+	}
+
+	return &AggregatedConditions[Obj]{
+		conditions: conditions,
+		state:      make([]*bool, len(conditions)),
+	}
+}
+
+func (ac AggregatedConditions[Obj]) clearState() {
+	for i := range ac.state {
+		ac.state[i] = nil
+	}
+}
+
+func (ac AggregatedConditions[Obj]) Condition(obj Obj) (bool, error) {
+	ac.clearState()
+
+	allDone := true
+	var err error
+	for i, cond := range ac.conditions {
+		var done bool
+		done, err = cond(obj)
+		ac.state[i] = &done
```

Comment:
shouldn't the err be checked before assigning the value? I think it would be more intuitive that you'd end up with a `<nil>` string for the condition that couldn't be evaluated, instead of a default value

Reply:
> for the condition that couldn't be evaluated

it **was** evaluated, ended with an error and this is its return value

nil is for the conditions that were not evaluated

Reply:
> and this is its return value

which, using this error pattern, you wouldn't normally take into account, since it returned an error

> nil is for the conditions that were not evaluated

I know what it is for now, but I'm saying imo it would make more sense to use it for the ones that returned an error, continue evaluation of other conditions and aggregate the errors, but I can live with the argument that it maintains the current behaviour

Reply:
>  you wouldn't normally take into account

You may or may not take it into account depending on function behaviour / docs, but it's the value that was returned which is my reasoning for including it. If the function meant to return no value, it would return `*bool`. Also with the done, the function return value is valid - anything that returns an error needs to return false.